grapher/file_grapher.py

```diff
@@ -0,0 +1,368 @@
+    Ref = namedtuple('Ref', ['DefPath', 'DefFile', 'Def', 'File', 'Start', 'End', 'ToBuiltin'])
+
+    _exported_regex = re.compile('\_[a-zA-Z0-9]')
+
+    def __init__(self, base_dir, source_file, log):
+        """
+        Create a new grapher.
+        """
+        self._base_dir = base_dir
+        self._file = source_file
+        # TODO: Can easily remove `log` to have cleaner class. Only used once.
+        self._log = log
+        self._source = None
+        self._defs = {}
+        self._refs = {}
+
+    def graph(self):
+        # TODO: Maybe move to `__init__`?
+        self._load()
+
+        # Add module/package defs.
+        module = os.path.splitext(self._file)[0]
+        if os.path.basename(self._file) == '__init__.py':
+            module = os.path.normpath(os.path.dirname(self._file))
+
+        self._add_def(self.Def(
+            Path=module,
+            Kind='module',
+            Name=module.split('/')[-1],
+            File=self._file,
+            DefStart=0,
+            DefEnd=0,
+            Exported=True,
+            # TODO: extract module/package-level doc.
+            Docstring='',
+            Data=None,
+        ))
+
+        # Jedi.
+        try:
+            jedi_names = jedi.names(source=self._source, path=self._file, all_scopes=True, references=True)
+        except Exception as e:
+            raise FileGrapherException('failed to parse {}: {}'.format(self._file, str(e)))
+
+        jedi_defs, jedi_refs = [], []
+        for jedi_name in jedi_names:
+            # Imports should be refs.
+            if jedi_name.is_definition() and jedi_name.type != 'import':
+                    jedi_defs.append(jedi_name)
+            else:
+                jedi_refs.append(jedi_name)
+
+        # Defs.
+        for jedi_def in jedi_defs:
+            self._log.debug(
+                '\nProcessing def: %s | %s | %s',
+                jedi_def.desc_with_module,
+                jedi_def.name,
+                jedi_def.type,
+            )
+            try:
+                self._add_def(self._jedi_def_to_def(jedi_def))
+            except Exception as e:
+                self._log.error(
+                    u'\nFailed to process def `%s`: %s',
+                    jedi_def.name,
+                    e,
+                )
+                continue
+
+        # Refs.
+        for jedi_ref in jedi_refs:
+            self._log.debug(
+                '\nProcessing ref: %s | %s | %s',
+                jedi_ref.desc_with_module,
+                jedi_ref.name,
+                jedi_ref.type,
+            )
+
+            # Try to find definition.
+            ref_def = jedi_ref
+            # If def is import, then follow it.
+            depth = 0
+            while (not ref_def.is_definition() or ref_def.type == "import") and depth < 100:
+                depth += 1
+                # noinspection PyBroadException
+                try:
+                    ref_defs = ref_def.goto_assignments()
```

Comment:
It does cache. I added that `depth` check not because of speed but because it can in some cases get into a loop.

I'm quite sure that I I add back multiple workers it will go down to 15 mins at least.